### PR TITLE
`Programming exercises`: Fix starting exercise after due date was postponed

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Sometimes instructors have to postpone a due date for various reasons. But after the due date some students might have already started the practice mode.
Starting the exercise after this scenario did not work.

### Description

Fix the bug

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise

1. Log in to Artemis
2. Set the due date of the exercise in the past
3. Participate using the practice mode
4. Set the due date into the future
5. Try to participate using the normal way
6. Set the due date in the past
7. Try to use the practice mode again

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ParticipationService.java | 77% | 93% |
| ParticipationResource | 63% | 90% |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 1]
```diff
@@ -323,12 +323,22 @@ public ResponseEntity<ProgrammingExerciseStudentParticipation> requestFeedback(@
         return ResponseEntity.ok().body(participation);
     }
 
-    private boolean isNotAllowedToStartProgrammingExercise(ProgrammingExercise programmingExercise, @Nullable StudentParticipation participation) {
+    /**
+     * Checks if the student is currently allowed to participate in the exercise using this Participation
+     * @param programmingExercise the exercise where the userr wants to participate
```
Comment (manuelmanso) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:328]:
```suggestion
     * @param programmingExercise the exercise where the user wants to participate
```

**Review by manuelmanso — APPROVED** [review]
Tested in ts1 and fix works. Code also looks good

**Review by stefanwaldhauser — DISMISSED** [review]
Tested it on TS1. Fix works. Code and Tests look good to! Gj 🎃

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 2]
```diff
@@ -329,9 +329,22 @@ public ResponseEntity<ProgrammingExerciseStudentParticipation> requestFeedback(@
...
+    private boolean isAllowedToParticipateInProgrammingExercise(ProgrammingExercise programmingExercise, @Nullable StudentParticipation participation) {
+        boolean isAfterDueDate = participation != null ? exerciseDateService.isAfterDueDate(participation)
                 : (programmingExercise.getDueDate() != null && now().isAfter(programmingExercise.getDueDate()));
+        if (participation != null) {
+            // Test runs may only be started after the due date
+            return participation.isTestRun() == isAfterDueDate;
+        }
+        else {
+            return !isAfterDueDate;
+        }
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:347]:
```suggestion
        if (participation != null) {
            // only regular participation before the due date; only practice run afterwards
            return participation.isTestRun() == exerciseDateService.isAfterDueDate(participation);
        }
        else {
            return programmingExercise.getDueDate() == null || now().isBefore(programmingExercise.getDueDate());
        }
```
I think the `?:` can be avoided here as the check for the `if` is the same to make the code more readable.

Please double-check that the suggestion does actually implement the same logic.

**Review by b-fein — DISMISSED** [review]
Code changes look good, only one suggestion where I think the readability can be improved with a small change. :+1:

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java (hunk 2 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java]: Your version looks better and works as well

**Review by manuelmanso — APPROVED** [review]
Reapprove!

**Review by Strohgelaender — APPROVED** [review]
code looks good.

**Review by b-fein — APPROVED** [review]
Code re-approve

**Review by simon-hng — APPROVED** [review]
Code changes look good

**Pull request merged**